### PR TITLE
[IMP] {l10n_in}_pos,point_of_sale: address missing dialog

### PR DESCRIPTION
**Pull request by meji-odoo**
Before this commit:
========================
 - Users are restricted from onboarding in the POS if the company address(state) is missing.

After this commit:
========================
 - If the company address(state) is missing, the popup message should appear when closing the session.
 - Users are allowed to enter the session even if the company address(state) is missing.
 - When closing the session, a redirect popup will be displayed.
 - Clicking on "Go to Company Configuration" in the redirect popup will automatically redirect user to the company page.

task-3792000

### CODE REVIEW

**Review by paso-odoo — CHANGES_REQUESTED** [review]
@meji-odoo 
Thanks for working on this PR.
Here are some small improvement points. 

Cheers!

File: addons/l10n_in_pos/static/src/overrides/models/closing_session.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { patch } from "@web/core/utils/patch";
+import { ConfirmationDialog } from "@web/core/confirmation_dialog/confirmation_dialog";
+
+patch(ClosePosPopup.prototype, {
+    async confirm() {
+        if (this.pos.company.country_id?.code === "IN" && !this.pos.company.state_id) {
+            this.dialog.add(ConfirmationDialog, {
+                title: _t("Company address is missing"),
+                body: _t(
+                    "Your company %s needs to have a correct address in order validate this invoice.\n" +
```
Comment (davidmonnom) [inline addons/l10n_in_pos/static/src/overrides/models/closing_session.js:15]: Validate this invoice ? Weird message when closing the session

File: addons/l10n_in/models/account_invoice.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def _post(self, soft=True):
         for move in posted.filtered(lambda m: m.country_code == 'IN'):
             if move.l10n_in_state_id and not move.l10n_in_state_id.l10n_in_tin:
                 raise UserError(_("Please set a valid TIN Number on the Place of Supply %s", move.l10n_in_state_id.name))
-            if not move.company_id.state_id:
+            if not (move.company_id.state_id or self._context.get('is_pos')):
```
Comment (davidmonnom) [inline addons/l10n_in/models/account_invoice.py:90]: I don't really understand the context stuff ? Is it really needed ?

**Review by davidmonnom — COMMENTED** [review]
Btw i don't know if its normal or not but you can create an invoice even without state_id set.

**Review by RicGR98 — COMMENTED** [review]
This is just delaying the error to the JS side instead of handling it right in the beginning on the python side... The onboarding is not really simplified if you still get the error but at a later step. I'm not a fan of going back-and-forth between the company settings and the POS... I feel like it'd be worth rechallenging the spec.

I don't understand the `is_pos` context key either, perhaps you should clarify its name ?

File: addons/l10n_in_pos/static/src/overrides/models/closing_session.js (hunk 1 above)
Comment (meji-odoo) [reply addons/l10n_in_pos/static/src/overrides/models/closing_session.js]: After discussion with PO. This message is uses in multiple times.

File: addons/l10n_in/models/account_invoice.py (hunk 2 above)
Comment (meji-odoo) [reply addons/l10n_in/models/account_invoice.py]: Remove context.

**Comment by meji-odoo** [comment]
@davidmonnom 
In Indian localization state_id is require for the tax calculation.

**Comment by meji-odoo** [comment]
> This is just delaying the error to the JS side instead of handling it right in the beginning on the python side... The onboarding is not really simplified if you still get the error but at a later step. I'm not a fan of going back-and-forth between the company settings and the POS... I feel like it'd be worth rechallenging the spec.
> 
> I don't understand the `is_pos` context key either, perhaps you should clarify its name ?

I have updated the code.
For the functional part please contact @mog-odoo.

File: addons/l10n_in_pos/static/src/compnay_state_dialog/compnay_state_dialog.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    static props = {
+        close: Function,
+    };
+
+    setup() {
+        this.pos = usePos();
+    }
+
+    redirect() {
+        window.location = "/odoo/settings/4/action-52/" + this.pos.company.id;
```
Comment (davidmonnom) [inline addons/l10n_in_pos/static/src/compnay_state_dialog/compnay_state_dialog.js:19]: I don't think the action identifier will be the same in all DBs.

**Review by davidmonnom — CHANGES_REQUESTED** [review]
To be honest this specs is very confusing for me. Is better in all way to put the popup at the opening and do not allow it.

File: addons/l10n_in_pos/static/src/compnay_state_dialog/compnay_state_dialog.js (hunk 3 above)
Comment (meji-odoo) [reply addons/l10n_in_pos/static/src/compnay_state_dialog/compnay_state_dialog.js]: I have updated the code.

**Comment by meji-odoo** [comment]
> To be honest this specs is very confusing for me. Is better in all way to put the popup at the opening and do not allow it.

@davidmonnom 
This task aims to enhance the user experience, as many POS users do not utilize the invoice section. Therefore, we cannot restrict them from accessing the POS, based on my understanding.
For the spec. related query please contact PO on task.

File: addons/l10n_in_pos/static/src/overrides/models/cash_move_popup.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+/** @odoo-module */
+
+import { CashMovePopup } from "@point_of_sale/app/navbar/cash_move_popup/cash_move_popup";
+
+import { patch } from "@web/core/utils/patch";
+import { companyStateDialog } from "../../compnay_state_dialog/compnay_state_dialog";
```
Comment (davidmonnom) [inline addons/l10n_in_pos/static/src/overrides/models/cash_move_popup.js:6]:
```suggestion
import { companyStateDialog } from "@l10n_in_pos/company_state_dialog/company_state_dialog";
```

File: addons/l10n_in_pos/static/src/overrides/models/closing_session.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+/** @odoo-module */
+
+import { ClosePosPopup } from "@point_of_sale/app/navbar/closing_popup/closing_popup";
+
+import { patch } from "@web/core/utils/patch";
+import { companyStateDialog } from "../../compnay_state_dialog/compnay_state_dialog";
```
Comment (davidmonnom) [inline addons/l10n_in_pos/static/src/overrides/models/closing_session.js:6]:
```suggestion
import { companyStateDialog } from "@l10n_in_pos/company_state_dialog/company_state_dialog";
```

File: addons/l10n_in_pos/static/src/compnay_state_dialog/compnay_state_dialog.js
Comment (davidmonnom) [inline addons/l10n_in_pos/static/src/compnay_state_dialog/compnay_state_dialog.js:1]: Typo on file name `compnay_state_dialog.js` -> `company_state_dialog.js`

**Review by davidmonnom — CHANGES_REQUESTED** [review]

File: addons/l10n_in_pos/static/src/overrides/models/pos_order.js
Comment (davidmonnom) [inline addons/l10n_in_pos/static/src/overrides/models/pos_order.js:1]: Line 44 of this file

```
baseLines.push({
    l10n_in_hsn_code: hsnCode,
    price_unit: priceUnit,
    quantity: line.get_quantity(),
    uom: null,
    ...getTaxesValues(
        taxes,
        priceUnit,
        1,
        line.product_id,
        this.company._product_default_values,
        this.company,
        this.currency
    ),
});
```

**Review by davidmonnom — CHANGES_REQUESTED** [review]
We're almost there, one last change and we should be good to go

File: addons/l10n_in_pos/static/src/overrides/models/pos_order.js
Comment (meji-odoo) [reply addons/l10n_in_pos/static/src/overrides/models/pos_order.js:1]: @davidmonnom 
Changes done.

**Comment by meji-odoo** [comment]
@odoo/rd-security 
Need to override security.

Thanks ...

**Comment by davidmonnom** [comment]
@odoo/rd-security can you check is just about a window.location.

Ty :)

**Comment by davidmonnom** [comment]
> > Need to override security.
> > Thanks ...
> 
> @odoo/pos need to explain why this is necessary and safe
> 
> Thanks ...

This is necessary because of the specification requirement, before we wouldn't allow pos to be launched without a state_id if we were in an Indian company. 

Now the check is done at the session closing, and in case there is no state_id, we display a popup asking if the user wants to be redirected to the company form view to add one.

@meji-odoo But come to think of it, shouldn't the redirection proposal only be made to users who have the right to edit the company?

Anyway, about `window.location`, it's OK because we're redirecting to an Odoo endpoint with an ID that comes from the server during the initial data loading.

**Comment by davidmonnom** [comment]
@meji-odoo can you rebase ?

**Comment by meji-odoo** [comment]
@davidmonnom 

I have squashed the commit.

Thank you.

**Comment by davidmonnom** [comment]
I've just tried the PR again and rebase on master and now when I try to launch the PoS I get this error

image

Is that normal ?

**Comment by meji-odoo** [comment]
@davidmonnom,

No, this isn't typical.
However, I didn't encounter this error in my branch after the rebase.
Below is the correct flow of this task.

https://github.com/odoo/odoo/assets/157008999/0e0f602d-f075-4074-8399-18c7961ffc13

Thank you..

**Comment by davidmonnom** [comment]
@robodoo r+

**Pull request closed without merging**